src/sdlsound.cpp

```diff
@@ -66,10 +66,193 @@ struct sound_effect {
+{
+    int adjusted = value + 1;
+    if( adjusted >= static_cast<int>( sfx_in_or_out::COUNT ) || adjusted < 0 ) {
+        throw std::invalid_argument( std::string( "sfx specified unknown inside/outside value " ) +
+                                     std::to_string( value ) );
+    }
+    return static_cast<sfx_in_or_out>( adjusted );
+}
+
+enum class sfx_time_of_day : uint8_t {
+    ANY = 0,
+    DAYTIME,
+    NIGHTTIME,
+    COUNT,
+};
+
+// This is encoded as an optional bool in json, so we cheat a little and accept -1 for 'not set'
+sfx_time_of_day tod_from_int( int value )
+{
+    int adjusted = value + 1;
+    if( adjusted >= static_cast<int>( sfx_time_of_day::COUNT ) || adjusted < 0 ) {
+        throw std::invalid_argument( std::string( "sfx specified unknown day/night value " ) +
+                                     std::to_string( value ) );
+    }
+    return static_cast<sfx_time_of_day>( adjusted );
+}
+
+// Fun but ugly template time.
+template<typename Map, typename Key>
+const std::vector<sound_effect> *find_sfx( const Map &c, Key &&k )
+{
+    auto it = c.find( std::forward<Key>( k ) );
+    if( it == c.end() ) {
+        return nullptr;
+    }
+    return &it->second;
+}
+
+template<typename Map, typename Key1, typename Key2, typename ...Keys>
+const std::vector<sound_effect> *find_sfx( const Map &c, Key1 &&k, Key2 &&k2, Keys &&...keys )
+{
+    auto it = c.find( std::forward<Key1>( k ) );
+    if( it == c.end() ) {
+        return nullptr;
+    }
+    return find_sfx( it->second, std::forward<Key2>( k2 ),
+                     std::forward<Keys>( keys )... );
+}
+
+template<typename Map, typename Key, typename Default>
+const std::vector<sound_effect> *find_closest_sfx( const Map &c, Key &&k, Default &&d )
+{
+    auto it = c.find( std::forward<Key>( k ) );
+    if( it == c.end() ) {
+        it = c.find( std::forward<Default>( d ) );
+    }
+    if( it == c.end() ) {
+        return nullptr;
+    }
+    return &it->second;
+}
+
+template<typename Map, typename Key1, typename Default1, typename Key2, typename Default2, typename ...KDs>
+const std::vector<sound_effect> *find_closest_sfx( const Map &c, Key1 &&k, Default1 &&d1, Key2 &&k2,
+        Default2 &&d2, KDs &&...kds )
+{
+    auto it = c.find( std::forward<Key1>( k ) );
+    if( it == c.end() ) {
+        it = c.find( std::forward<Default1>( d1 ) );
+    }
+    if( it == c.end() ) {
+        return nullptr;
+    }
+    return find_closest_sfx( it->second, std::forward<Key2>( k2 ),
+                             std::forward<Default2>( d2 ),
+                             std::forward<KDs>( kds )... );
+
+}
+
+template<typename Map, typename Key>
+std::vector<sound_effect> &emplace_sfx( Map &c, Key &&k )
+{
+    return c[std::forward<Key>( k )];
+}
+
+template<typename Map, typename Key1, typename Key2, typename ...Keys>
+std::vector<sound_effect> &emplace_sfx( Map &c, Key1 &&k, Key2 &&k2, Keys &&...keys )
+{
+    auto &nested_container = c[std::forward<Key1>( k )];
```

Comment:
This was `auto` like all the ones above it, but that caused a use-after-free in the loading code. D'oh.